src/shapes.nim

```diff
@@ -0,0 +1,277 @@
+from std/fenv import epsilon
+from std/math import sgn, floor, sqrt, arccos, arctan2, PI
+import std/options
+
+import geometry, camera
+
+
+type
+    AABB* = tuple[min, max: Point3D]
+
+    ShapeKind* = enum
+        skAABox, skTriangle, skSphere, skPlane, skMesh
+
+    Shape* = object of RootObj
+        transf*: Transformation
+        aabb*: Option[AABB] = none(AABB)
+
+        case kind*: ShapeKind 
+        of skAABox: 
+            min, max: Point3D
+
+        of skTriangle: 
+            vertices*: tuple[A, B, C: Point3D]
+
+        of skMesh:
+            nodes*: seq[Point3D]
+            edges*: seq[tuple[left, right: int]]
+
+        of skSphere:
+            center*: Point3D
+            radius*: float32
```

Comment:
I feel like we could do a little better here: old code enabled us to create a rotation ellipsoid just by applying a non isotropic scaling and it was still a Sphere type variable. Here we are defining some features which are typical of a sphere like center and radius and a new Sphere is given by the following procs:
``` nim
proc newSphere*(center: Point3D, radius: float32): Shape {.inline.} = 
    Shape(
        kind: skSphere,
        transf: newTranslation(center.Vec3f) @ newScaling(radius), 
        center: center, radius: radius
    )
```
We are giving as inputs center and radius, which is something i like but i would like to also have a kind like ```Ellipsoid``` which has:
``` nim
Shape* = object of RootObj
    transf*: Transformation
    aabb*: Option[AABB] = none(AABB)

    case kind*: ShapeKind
    of skEllipsoid:
        center*: Point3D
        a*, b*, c*: float32 (which are floating point number regarding semi-axles )
        theta*, phi*: float32 (which are angles to orient the shape as we desire)
```

Reply:
We should also add Cylinder and Cube